scvi/dataset/dataset.py

```diff
@@ -1214,6 +1228,101 @@ def raw_counts_properties(
+        batch_correction: Optional[bool] = True,
+    ) -> pd.DataFrame:
+        """\
+        Code sample taken from the scanpy package
+        Annotate highly variable genes [Satija15]_ [Zheng17]_.
+        Depending on `flavor`, this reproduces the R-implementations of Seurat
+        [Satija15]_ and Cell Ranger [Zheng17]_.
+        The normalized dispersion is obtained by scaling with the mean and standard
+        deviation of the dispersions for genes falling into a given bin for mean
+        expression of genes. This means that for each bin of mean expression, highly
+        variable genes are selected.
+        Parameters
+        ----------
+        :param min_mean:
+            If `n_top_genes` unequals `None`, this and all other cutoffs for the means and the
+            normalized dispersions are ignored.
+        :param max_mean:
+            If `n_top_genes` unequals `None`, this and all other cutoffs for the means and the
+            normalized dispersions are ignored.
+        :param min_disp:
+            If `n_top_genes` unequals `None`, this and all other cutoffs for the means and the
+            normalized dispersions are ignored.
+        :param max_disp:
+            If `n_top_genes` unequals `None`, this and all other cutoffs for the means and the
+            normalized dispersions are ignored.
+        :param n_top_genes:
+            Number of highly-variable genes to keep.
+        :param n_bins:
+            Number of bins for binning the mean gene expression. Normalization is
+            done with respect to each bin. If just a single gene falls into a bin,
+            the normalized dispersion is artificially set to 1. You'll be informed
+            about this if you set `settings.verbosity = 4`.
+        :param flavor:
+            Choose the flavor for computing normalized dispersion. In their default
+            workflows, Seurat passes the cutoffs whereas Cell Ranger passes
+            `n_top_genes`.
+        :param batch_correction:
+            Whether batches should be taken into account during procedure
+
+        :return:
+            scanpy .var DataFrame providing genes information including means, dispersions
+            and whether the gene is tagged highly variable (key `highly_variable`)
+            (see scanpy highly_variable_genes documentation)
+
+        """
+
+        logger.info("extracting highly variable genes")
+        try:
+            import scanpy as sc
+        except ImportError:
+            raise ImportError(
+                "please install scanpy: " "pip install scanpy python-igraph louvain"
+            )
+
+        # Creating AnnData structure
+        obs = pd.DataFrame(
+            data=dict(batch=self.batch_indices.squeeze()),
+            index=np.arange(self.nb_cells),
+        ).astype("category")
+
+        counts = self.X
+        if issparse(counts):
+            counts = counts.toarray()
+        adata = sc.AnnData(X=counts, obs=obs)
```

Comment:
does the `obs` you create earlier, does it definitely have the name "batch"?

Reply:
you could also do `adata.obs["batch"] = self.batch_indices.squeeze()`

Reply:
`data=dict(batch=self.batch_indices.squeeze())` definitely adds the feature `batch` to `obs`